### PR TITLE
Log entry coloring.

### PR DESCRIPTION
**Pull request by pq**
Basic entry coloring (WIP).

image

As discussed in #2200, this will ultimately benefit from user configurability; this plumbs in the back-end for that.

/cc @devoncarew @scheglov

### CODE REVIEW

**Comment by quangson91** [comment]
@pq, how did you get level `WARNING` in the console ?
I've tried:
```dart
  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen((LogRecord rec) {
    print('${rec.level.name}: ${rec.time}: ${rec.message}');
  });

  log.finest("LOG FINEST");
  log.finer("LOG FINER");
  log.fine("LOG FINE");
  log.config("LOG CONFIG");
  log.info("LOG INFO");
  log.warning("LOG WARNING");
  log.severe("LOG SEVERE");
  log.shout("LOG SHOUT");
```
But the log level all show `INFO`

File: src/io/flutter/logging/FlutterLogView.java [hunk 1]
```diff
@@ -39,8 +37,32 @@
 import java.awt.datatransfer.StringSelection;
 import java.awt.event.MouseEvent;
 
+import static com.intellij.ui.SimpleTextAttributes.STYLE_PLAIN;
+
 public class FlutterLogView extends JPanel implements ConsoleView, DataProvider, FlutterLog.Listener {
 
+  private class EntryModel implements FlutterLogTree.EntryModel {
+    boolean color;
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogView.java]: we should rename that variable. Eg, `isEnableLogColor` or something like that.
(`color` feels like the actual color value - Colors.RED)

File: src/io/flutter/logging/FlutterLogView.java [hunk 2]
```diff
@@ -39,8 +37,32 @@
...
+
+    @Override
+    public SimpleTextAttributes style(@Nullable FlutterLogEntry entry, int attributes) {
+      if (color && entry != null) {
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogView.java]: I'm not sure about the java coding style.
But I think we can use earlier return to reduce the deep nested.
Eg
```java
if (entry == null || notEnableLogColor) return;
// Do other things.
```

File: src/io/flutter/logging/FlutterLogView.java [hunk 3]
```diff
@@ -39,8 +37,32 @@
...
+        final FlutterLog.Level level = FlutterLog.Level.forValue(entry.getLevel());
+        if (level != null) {
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogView.java]: earlier return too.

File: src/io/flutter/logging/FlutterLogView.java [hunk 4]
```diff
@@ -300,8 +342,10 @@ String countString(int count) {
       return Integer.toString(count);
     }
   }
-
   @NotNull final FlutterApp app;
+  private final SimpleTextAttributes ORANGE_ATTRIBUTES = new SimpleTextAttributes(STYLE_PLAIN, JBColor.orange);
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogView.java]: Maybe add annotation `@NotNull`

**Comment by pq** [comment]
Hey @quangson91.  To get that to work, you'll want to plumb level through to the `developer` log.  Try replacing your `print` with something like this:

```dart
  Logger.root.onRecord.listen((LogRecord rec) {
    developer.log(rec.message, level: rec.level.value);
  });
```

with an added import:

```dart
import 'dart:developer' as developer;
```


Hope that helps!

**Comment by quangson91** [comment]
> Logger.root.onRecord.listen((LogRecord rec) {
>    developer.log(rec.message, level: rec.level.value);
> });

@pq it worked.
Thank you!

File: src/io/flutter/logging/FlutterLogView.java [hunk 5]
```diff
@@ -39,8 +37,32 @@
 import java.awt.datatransfer.StringSelection;
 import java.awt.event.MouseEvent;
 
+import static com.intellij.ui.SimpleTextAttributes.STYLE_PLAIN;
```
Comment (scheglov) [inline src/io/flutter/logging/FlutterLogView.java]: We later use other values of SimpleTextAttributes.
Should we statically import all of them, or none?

Comment (pq) [reply src/io/flutter/logging/FlutterLogView.java]: Good catch.  Thanks!

File: src/io/flutter/logging/FlutterLogView.java (hunk 1 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogView.java]: In a private classes I prefer shorter names, especially when the meaning should be clear from the context (and the type).  Of course, the fact that it surprised you, means it wasn't clear! 😄  I'll make it better.

Thanks!

File: src/io/flutter/logging/FlutterLogView.java (hunk 2 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogView.java]: I'm on the fence.  Many argue that having multiple return points in a function is as confusing (if not more). 🤔 

That said, when a function is very short and can be read in a glance I tend to think either approach is fine.

Thanks!

File: src/io/flutter/logging/FlutterLogView.java (hunk 3 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogView.java]: Cool.  Thanks.  See above for my rationale.

File: src/io/flutter/logging/FlutterLogView.java (hunk 4 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogView.java]: 👍

**Pull request merged**